[Core][Geometry] Brep geometry classes

tteschemacher opened the pull request:
Ok, so let's begin the game... This PR is not ready to merge, yet.

I like to discuss here about a geometry type, which is actually acting as a single integration point. All necessary values as shape functions (and any amount of derivatives), integration weight, and its nodes, are computed in advance and set to it.

This type of geometry could be used for any discretization and type of shape function, as e.g. NURBS/B-Splines or classical linear FEM-shape functions. Typically, it is nice to use it for elements that only have one GP as MPM, but also in the current design of IBRA/IGA this would ease and clean up some procedures.

In this desgin the geometry.h would have to be touched slightly. This geometry is not based on GeometryData, as everything is preset. Thus, many functions from the geometry would have to be set to virtual (as e.g. Dimension, WorkingSpaceDimension, LocalSpaceDimension, ShapeFunctionValue, ShapeFunctionValues, including all derivatives, ...) because the derived object would need them different.

Additionally, to this functions as Normal/UnitNormal should be generalized to have also the function call with a IntegrationPointIndex.

Also, I think we should discuss about a generalization of the shape function and shape function derivatives getter functions. For some cases it would be nice to give a call and just pass the order of derivative as int. In IGA we could give a unlimited amount of derivatives.

tteschemacher committed:
Integration Point Surface 3D Geometry and registration

tteschemacher committed:
New integration_point_surface_3d.h for single integration points with registration.

tteschemacher committed:
making functions virtual

loumalouomega commented on kratos/geometries/geometry.h:
```diff
@@ -540,7 +540,7 @@ class Geometry : public PointerVector<TPointType>
     @see LocalSpaceDimension()
     */
-    inline SizeType Dimension() const
+    virtual inline SizeType Dimension() const
```
This is enough to block just in case

tteschemacher replied:
Sure, as mentioned for the moment just for discussion. It doesnt compile like this.

loumalouomega reviewed (changes requested):
Blocking for the moment until someone from @KratosMultiphysics/technical-committee gives green light

pooyan-dadvand commented:
I'm in the game!!

First question, why you dropped the GeometryData? I understand that you cannot fill/use all of its data but there many (like Diemensions) can be used and making virtual such small methods would affect the performance. So I would suggest to separate the methods to become virtual not due to the GeometryData but due to the fact that you can not pre-calculate them.

tteschemacher committed:
Changes to use GeometryData members with slight changes in the geometry_data.

philbucher reviewed (commented):
now without changes to `geometry.h` :+1:

tteschemacher commented:
Ok. From first look it was not very intuitive to use those, but the advantages are definiteley given. I modified some functions in the geometry_data (I added some set functions, which only influence compile time). With those modifications geometry does not has to be touched anymore, thus no efficiency problems.

tteschemacher committed:
Merge branch 'master' into feature/integration_point_surface_3d_geometry

tteschemacher committed:
Adding msGeometryData member

pooyan-dadvand commented on kratos/geometries/integration_point_surface_3d.h:
```diff
@@ -0,0 +1,473 @@
+                ShapeFunctionsLocalGradientsIntegrationPoint(i, 1) = rShapeFunctions(2, i);
+            }
+            ShapeFunctionsLocalGradients.push_back(ShapeFunctionsLocalGradientsIntegrationPoint);
+            msGeometryData.ShapeFunctionsLocalGradients(ShapeFunctionsLocalGradients);
```
Are you sure that you want to change the Geometry Data for all objects of type `IntegrationPointSurface3d`? 

If the number of integration points depends to the number of points you cannot use the geometrical data and we should make the correspon methods virtual

tteschemacher replied:
Actually, I don't see a blocker and for all other cases it is more efficient.

I see GeometryData only as data container. E.g. in Iga we don't follow standard integration rules (especially for trimmed geometries). Thus, I want a possibility to fill up the container with data which is computed somewhere but not necessariliy at construction time and not from inside the geometry (The reason is it should be independent on shape function. NURBS/ linear FEM/ whatever...). Those setters are the only functions with which I added to the GeometryData. With those, geometry.h does not has to be touched.

Anyways, please see this IntegrationPointGeometry as something special. It shall be the latest entitity in the chain. Due to this generity it can be used anywhere. This is nice for e.g. MPM (change the background whenever to whatever - the shape function type is defined), in IGA many different ways to discretize surface have their validity and are done from outside and latest this was not discussed fully yet, but for interfaces between physics this can be also used, with this we would have a standardized form anywhere.

Maybe this figure describes it a bit better:
![image](https://user-images.githubusercontent.com/25983517/56802579-b23b8c80-6820-11e9-8f56-334d2ca7e633.png)


Later the BrepGeometries might not even need this modification in the GeometryData. But, let's see this.

pooyan-dadvand replied:
My question are all surfaces share the same gradient array in entire Kratos?

tteschemacher replied:
Actually, first, I didn't understand your issue. Now I do. And you're right, we cannot handle this with this static GeometryData member.

Gradients and Shape Functions are different for each of those geometries. But, we can keep at least dimensions, ...

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -407,6 +416,15 @@ class GeometryData
+    IntegrationPointsArrayType which is Vector of integration points
+    for default integrating method.
+    */
+    void IntegrationPoints(IntegrationPointsArrayType& IntegrationPoints)
```
```suggestion
    void SetIntegrationPoints(const IntegrationPointsArrayType& IntegrationPoints)
```

tteschemacher replied:
Thanks for the comment. I can live with both options.

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -438,6 +456,21 @@ class GeometryData
+    node j evaluated in integration point i of default integration method.
+
+    */
+    void ShapeFunctionsValues(Matrix& ShapeFunctionsValues)
```
```suggestion
    void SetShapeFunctionsValues(const Matrix& ShapeFunctionsValues)
```

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -552,6 +585,21 @@ class GeometryData
+    the shape function corresponding to node j evaluated in
+    integration point i of default integration method.
+    */
+    void ShapeFunctionsLocalGradients(ShapeFunctionsGradientsType& ShapeFunctionsLocalGradients)
```
```suggestion
    void SetShapeFunctionsLocalGradients(const ShapeFunctionsGradientsType& ShapeFunctionsLocalGradients)
```

msandre commented on kratos/geometries/integration_point_surface_3d.h:
```diff
@@ -0,0 +1,473 @@
+        const PointsArrayType& ThisPoints,
+        const CoordinatesArrayType& rLocalCoordinates,
+        const double& rIntegrationWeight,
+        Matrix& rShapeFunctions)
```
This way of passing shape function values and derivatives is very prone to misuse and bugs imo. I would suggest cleaning this up if the overall idea is given a green light.

tteschemacher replied:
thanks for your issues. I know that this is a bit difficult and one would have to think precisely. But like this one can pass as many derivatives as wished. In high order this might be very important.

Maybe we can also bypass this with a nested vector.

tteschemacher replied:
A comment about the array dimensions: Through the standard construction they might have the same problem. The integration points, shape functions and derivatives are all computed in separate functions.

Maybe in the setters one could leave some checks. Or all values are passed in the same function and there are the dimension\ array length checks.

msandre replied:
> Or all values are passed in the same function and there are the dimension\ array length checks.

Agreed. Eg., 

```
void GeometryData::SetGeometryData(const IntegrationPointsContainerType& ThisIntegrationPoints,
                  const ShapeFunctionsValuesContainerType& ThisShapeFunctionsValues,
                  const ShapeFunctionsLocalGradientsContainerType& ThisShapeFunctionsLocalGradients );
```

msandre reviewed (commented):
There is a bit of a trade-off here wrt to the new setter member functions of `GeometryData`. On the one hand they add flexibility by allowing the user to vary members independently. On the other hand the members share an invariant (array dimensions) so that they are not really independent from one another. Thus the modified interface requires more knowledge (time) from the user not to violate the invariant. In some cases this could lead to undefined behavior, but I think with the current implementation one would at least get a seg fault if they try to access out of bounds. Still it's something to consider imo.

tteschemacher committed:
Independent Shape Function Order

msandre commented on kratos/geometries/geometry.h:
```diff
@@ -836,6 +860,54 @@ class Geometry : public PointerVector<TPointType>
+    /**
+    * @brief TO BE DONE
+    */
+    virtual array_1d<double, 3> Normal(array_1d<double, 3>& rResult, IndexType IntegrationPointIndex, IntegrationMethod ThisMethod) const
```
```suggestion
    array_1d<double, 3> Normal(const Geometry& rGeom, IndexType IntegrationPointIndex, IntegrationMethod ThisMethod); // free function
```
This function can go outside of the class since it doesn't need to access private data.

pooyan-dadvand replied:
I disagree.... It is better to be were data is and all data are stored in the geometry.

tteschemacher replied:
I just followed the structure of the functions for shape functions, gradients, ... . It is only for completeness. The other Normal function is right ahead.

msandre replied:
@pooyan-dadvand that's fine and I know this isn't the convention in Kratos. But it is considered good practice. For reasons see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-member and https://isocpp.org/blog/2018/02/free-your-functions-klaus-iglberger

pooyan-dadvand replied:
The rule applies if `normal` function doesn't depends on type of the geometry regardless to if data accessed or not

msandre replied:
I agree with that reasoning 👍 But we should be more careful with the other since it can really affect code quality.

msandre commented on kratos/geometries/geometry.h:
```diff
@@ -836,6 +860,54 @@ class Geometry : public PointerVector<TPointType>
+    /**
+    * @brief TO BE DONE
+    */
+    virtual array_1d<double, 3> Normal(array_1d<double, 3>& rResult, IndexType IntegrationPointIndex) const
```
same

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -310,6 +357,25 @@ class GeometryData
+        const IntegrationPointsContainerType& ThisIntegrationPoints,
+        const ShapeFunctionsDerivativesVectorType& ThisShapeFunctionsDerivativesVector)
+    {
+        for (SizeType i = 0; i < ThisShapeFunctionsDerivativesVector[mDefaultMethod].size(); ++i)
```
also check `ThisIntegrationPoints.size() == mShapeFunctionsValues[mDefaultMethod].size1()`

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -438,6 +504,17 @@ class GeometryData
+
+    const ShapeFunctionsDerivativesIntegrationPointsType& ShapeFunctionsDerivativesIntegrationPoints(IndexType DerivativeOrder) const
+    {
+        return mShapeFunctionsDerivativesContainer[mDefaultMethod][DerivativeOrder];
```
```suggestion
        return mShapeFunctionsDerivativesContainer[DerivativeOrder][mDefaultMethod];
```
defining the container this way would simplify the implementation. Then you could replace `mShapeFunctionsLocalGradients` with mShapeFunctionsDerivativesContainer[1];`

tteschemacher replied:
I think like this it could be the same. One never returns the container. Thus, would just be mShapeFunctionsLocalGradients = mShapeFunctionsLocalGradients[mDefaultMethod][1]. Anyways, I can live with both.

msandre replied:
I would not do `mShapeFunctionsLocalGradients = mShapeFunctionsLocalGradients[mDefaultMethod][1]` as it is wastes memory and you need to synchronize data.

msandre commented on kratos/geometries/integration_point_surface_3d.h:
```diff
@@ -0,0 +1,431 @@
+private:
+    ///@name Static Member Variables
+    ///@{
+    const GeometryData msGeometryData;
```
```suggestion
    GeometryData mGeometryData;
```

tteschemacher replied:
changed, thanks!

msandre commented on kratos/geometries/integration_point_surface_3d.h:
```diff
@@ -0,0 +1,431 @@
+    const GeometryData msGeometryData;
+
+    /** This member holds the higher order shape function derivatives */
+    Matrix mShapeFunctions;
```
```suggestion
```
I don't think this is needed.

tteschemacher replied:
you're right. I remove it.

msandre commented on kratos/geometries/geometry_data.h:
```diff
@@ -263,6 +296,20 @@ class GeometryData
+        : mDimension(ThisDimension)
+        , mWorkingSpaceDimension(ThisWorkingSpaceDimension)
+        , mLocalSpaceDimension(ThisLocalSpaceDimension)
+    {
```
```suggestion
        , mDefaultMethod()
    {
```

tteschemacher committed:
remove m ShapeFunctions

tteschemacher committed:
Small modifications towards integration point geometry

tteschemacher committed:
Introduction of brep_face.h.

tteschemacher committed:
updates. more functionalities.

tteschemacher committed:
Example Brep Curve

tteschemacher committed:
Nurbs edge on face

tteschemacher committed:
Coupling geometry, with container for multiple geometries

tteschemacher committed:
integration point for line on surface point

tteschemacher committed:
Possible design of coupling condition

tteschemacher committed:
updates in coupling condition

tteschemacher committed:
Updates in PenaltyCouplingCondition

tteschemacher committed:
Adding comments

tteschemacher committed:
Updates in CouplingMasterSlave. GetGeometryPart...

pooyan-dadvand commented on applications/IgaApplication/custom_condition/penalty_coupling_condition.cpp:
```diff
@@ -30,26 +30,40 @@ namespace Kratos
 
         const auto& r_geometry_master = GetGeometry();
-        const auto& r_geometry_slave = r_geometry_master.GetSlaveGeometry();
+        const auto& r_geometry_slave = r_geometry_master.GetGeometryPart(1);
```
You may define an enum in your master slave geometry to be more readable with something like:
```c++
const auto& r_geometry_slave = r_geometry_master.GetGeometryPart(MasterSlaveGeometry::SlavePart);
```

tteschemacher replied:
But this does not has the possibility to have more than one slave. I will think about something.

tteschemacher committed:
updates

tteschemacher commented on kratos/geometries/coupling_master_slave.h:
```diff
@@ -0,0 +1,452 @@
+        typename GeometryType::Pointer pMasterGeometry,
+        typename GeometryType::Pointer pSlaveGeometry,
+        IsGeometryDataInitialized = false)
+        : BaseType((*pMasterGeometry)(), &(pMasterGeometry->msGeometryData)),
```
@pooyan-dadvand do you see a possibility to pass the GeometryData of the master to the coupling geometry?

pooyan-dadvand replied:
I would add a `GeometryData const& GetGeometryData() const` method to base class

loumalouomega commented on kratos/geometries/coupling_master_slave.h:
```diff
@@ -0,0 +1,452 @@
+ *        to a surface geometry and a list of BrepEdge s. Those edges
+ *        define the trimming of the respective surface.
+ */
+template<class TPointType> class CouplingMasterSlave
```
@tteschemacher and @pooyan-dadvand  Just to mention that what I do for the ContactStructuralMechanicsApplication is define a coomon condition to all contact conditions is define a paired condition https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ContactStructuralMechanicsApplication/custom_conditions/paired_condition.h

I am not against this, but I would like to know exactly how this works?, this geometry contains N geometries?, being one of them the "reference condition" (master/slave depending of your definition)?

tteschemacher replied:
The idea is very similiar to yours. However it does not relate the conditions but directly the geomtries. Thus, we can use the standard constructors for conditions and elements. And don't have to do Set/GetValue operations.

The access GetGeometry() will provide the geometry of the Master. With GetGeometryPart(index) we can then access to any arbitrary part, genereally the slave. (this can handle more than one couplings, or also embedded sliding geometries, ...)

loumalouomega replied:
For me OK, that that would imply to refactor many things in my app, but I will not block as long as this makes completely sense

tteschemacher committed:
LagrangeCouplingCondition

tteschemacher committed:
Merge branch 'feature/integration_point_surface_3d_geometry' of https://github.com/KratosMultiphysics/Kratos into feature/integration_point_surface_3d_geometry

# Conflicts:
#	kratos/geometries/integration_point_surface_3d.h

philbucher commented:
do you still need this PR?

tteschemacher commented:
This PR is not needed anymore. Thanks all for the discussion! This PR was split up and was the basis for many developments.

philbucher commented:
do you still need the branch?

loumalouomega commented:
Is the brach still needed?
